File: python/day10.py
from utilities import read_int_input
from collections import defaultdict
from typing import List, Dict


def count_jolt_differences(adapters: List[int]) -> Dict[int, int]:
    differences = defaultdict(int)
    # difference of the device and the highest jolt adpater
    differences[3] = 1
    prev = 0
    for adapter in adapters:
        differences[adapter - prev] += 1
        prev = adapter
    return differences


def find_number_of_arrangements(adapters: List[int]) -> int:
    jolts = [0] + adapters
    arrangements = [0] * len(jolts)
    arrangements[0] = 1
    for i in range(1, len(arrangements)):
        for j in range(1, 4):
            if i - j > -1 and jolts[i] - 3 <= jolts[i - j]:
                arrangements[i] += arrangements[i - j]
    return arrangements[-1]


def part_1(adapters: List[int]) -> int:
    differences = count_jolt_differences(adapters)
    return differences[1] * differences[3]


def part_2(adapters: List[int]) -> int:
    return find_number_of_arrangements(adapters)


if __name__ == '__main__':
    print('Day 10')
    adapters = read_int_input('../inputs/day10.txt')
    adapters.sort()
    print('\tPart 1: {}'.format(part_1(adapters)))
    print('\tPart 2: {}'.format(part_2(adapters)))

File: rust/src/day10.rs
use crate::utilities::read_int_array;

fn find_differences_product(adapters: &Vec<i64>) -> i64 {
    let mut prev = 0;
    let mut diff_1 = 0;
    let mut diff_3 = 1;
    for &adapter in adapters {
        if adapter - prev == 1 {
            diff_1 += 1;
        } else if adapter - prev == 3 {
            diff_3 += 1;
        }
        prev = adapter;
    }
    diff_1 * diff_3
}

fn find_num_of_arrangments(adapters: &Vec<i64>) -> i64 {
    let mut arrangements = vec![0; adapters.len()];
    for i in 0..3 {
        if adapters[i] <= 3 {
            arrangements[i] = 1;
        }
    }
    for i in 1..adapters.len() {
        for j in 1..4 {
            if i >= j && adapters[i] - adapters[i - j] <= 3 {
                arrangements[i] += arrangements[i - j];
            }
        }
    }
    arrangements.pop().unwrap()
}

fn part_1(adapters: &Vec<i64>) {
    println!("\tPart 1: {}", find_differences_product(adapters));
}

fn part_2(adapters: &Vec<i64>) {
    println!("\tPart 2: {}", find_num_of_arrangments(adapters));
}

pub fn main() {
    println!("Day 10");
    let mut adapters = read_int_array("../inputs/day10.txt");
    adapters.sort();
    part_1(&adapters);
    part_2(&adapters);
}
